services/brig/src/Brig/Data/MLS/KeyPackage.hs

```diff
@@ -194,9 +195,47 @@ addKeyPackageRef ref nkpr = do
     q :: PrepQuery W (ClientId, ConvId, Domain, Domain, UserId, KeyPackageRef) x
     q = "UPDATE mls_key_package_refs SET client = ?, conv = ?, conv_domain = ?, domain = ?, user = ? WHERE ref = ?"
 
+-- | Update key package ref, used in Galley when commit reveals key package ref update for the sender.
+-- Nothing is changed if the previous key package ref is not found in the table.
+-- Updating amounts to INSERT the new key package ref, followed by DELETE the
+-- previous one.
+--
+-- FUTUREWORK: this function has to be extended if a table mapping (client,
+-- conversation) to key package ref is added, for instance, when implementing
+-- external delete proposals.
+updateKeyPackageRef :: MonadClient m => KeyPackageRef -> KeyPackageRef -> m ()
+updateKeyPackageRef prevRef newRef =
+  void . runMaybeT $
+    backupKeyPackageMeta prevRef
+      >>= ($) lift . ((>> deleteKeyPackage prevRef) . restoreKeyPackageMeta newRef)
+
 --------------------------------------------------------------------------------
 -- Utilities
 
+backupKeyPackageMeta :: MonadClient m => KeyPackageRef -> MaybeT m (ClientId, ConvId, Domain, Domain, UserId)
+backupKeyPackageMeta ref = do
+  MaybeT . retry x1 $ query1 q (params LocalQuorum (Identity ref))
+  where
+    q :: PrepQuery R (Identity KeyPackageRef) (ClientId, ConvId, Domain, Domain, UserId)
+    q = "SELECT client, conv, conv_domain, domain, user FROM mls_key_package_refs WHERE ref = ?"
+
+restoreKeyPackageMeta :: MonadClient m => KeyPackageRef -> (ClientId, ConvId, Domain, Domain, UserId) -> m ()
```

Comment:
I have the same note about the two occurrences of `Domain` in the tuple in the type signature as above.

Reply:
Done. See [ea33455](https://github.com/wireapp/wire-server/pull/2604/commits/ea334556b017d02bd9643cd89caeb041e78ccce0).